src/components/Location.react.js

```diff
@@ -1,39 +1,85 @@
+        } = props;
 
-        /*
-         * if pathname isn't defined, then it wasn't set by the user
-         * and it needs to be equal to the window.location.pathname
-         * this just happens on page load
+        /**
+         * Check if the field exists in props. If the prop with "fieldName" is not defined,
+         * then it was not set by the user and needs to be equal to the value in window.location.
+         * This only happens on page load (since props will no longer be undefined after componentDidMount).
+         *
+         * @param {string} fieldName
+         *  The name of the prop in window.location and in the component's prop
+         *
+         * @returns {boolean}
+         *  Returns true if the prop with fieldName is different and the window state needs to be updated
          */
-        if (!pathname) {
-            setProps({pathname: window.location.pathname});
-        } else if (pathname !== window.location.pathname) {
-            if (props.refresh) {
-                // Refresh the page
-                window.location.pathname = pathname;
-            } else {
-                // Just push the new location into the URL
-                if (setProps) setProps({pathname});
-                window.history.pushState({}, '', pathname);
+        const checkExistsUpdateWindowLocation = (fieldName) => {
+            const propVal = props[fieldName];
+
+            if (!propVal && window.location[fieldName] && setProps) { // Prop is undefined?
+                const update = {};
+                update[fieldName] = window.location[fieldName];
+                setProps(update);
+            } else if (propVal && (propVal !== window.location[fieldName])) { // Prop has changed?
+                if (refresh) { // Refresh the page?
+                    window.location[fieldName] = propVal;
+                } else if ((this.props[fieldName] !== propVal)) { // If this prop has changed, need to setProps
+                    const update = {};
+                    update[fieldName] = propVal;
+
+                    if (setProps)
+                        setProps(update);
+                    return true; // This needs to be pushed in the window.history
+                }
             }
+
+            return false;
+        };
+
+        // Check if the prop value needs to be updated (note that this mutates propsToSet)
+        const pathnameUpdated = checkExistsUpdateWindowLocation('pathname');
+        const hrefUpdated = checkExistsUpdateWindowLocation('href');
+        const hashUpdated = checkExistsUpdateWindowLocation('hash');
+        const searchUpdated = checkExistsUpdateWindowLocation('search');
+
+        if (hrefUpdated) // Special case -- overrides everything!
+            window.history.pushState({}, '', href);
+        else if (pathnameUpdated || hashUpdated || searchUpdated) { // Otherwise, we can mash everything together
+            const searchVal = typeof search !== 'undefined' ? search : '';
```

Comment:
for consistency, I use Ramda in the lib, so instead of `typeof search !== 'undefined'`, `R.Type(search) !== 'Undefined'`

Reply:
Great. Haven't used that before so appreciate the example here and the links to the documentation in another comment.